### DISCUSSION TITLE
Mem cost problem for Milvus performance benchmark

### DISCUSSION THREAD

**Discussion by simonwei97** [milvus-io/milvus#27746] [Q&A and General discussion] [answered]
I performed benchmark tool on standalone milvus, which version is `v2.2.13`.

Refer benchmark guide
- https://github.com/milvus-io/milvus-tools/blob/main/benchmark/README.md
- https://milvus.io/blog/2022-08-16-A-Quick-Guide-to-Benchmarking-Milvus-2-1.md

## Configure for Milvus
- Milvus version: `v2.2.13`
- Milvus deploymode: `standalone`
- Milvus resources: 
   - CPU: `Limits=32, Requests=4`
   - Mem: `Limits=32GiB, Requests=4GiB`

## Collection info
- collection schema:
   - uuid: uuid format, occupied `36 Byte`
   - xx_id: int64, occupied `8 Byte`
   - xx_uuid: uuid format, occupied `36 Byte`
   - type: string, occupied `4 Byte`
   - index: int64, occupied `8 Byte`
   - text: 100 chinese char, occupied `3*100 Byte`
   - vector: dim=768, occupied `4*768 Byte`
> for each entity in collection, it will occupied `3464 Byte`, approximately equal to `3.38 KB`.

- index parameter: `{'index_type': 'FLAT', 'metric_type': 'L2'}`

## Result

|  Total inserted entities  | mem cost (Total)  |     mem cost (single entity)     |
|  :-----------: |  :-----------: |  :-------------------------: |
|  10W            | 2.0831 GB       |  21.8429 KB |
|  50W            | 3.597 GB        |  7.5435 KB |
|  100W          | 4.8211 GB       |  5.0553 KB |
|  500W         | 22.0009 GB   |  4.6139 KB |
|  661w          | 30.7090 GB   |  4.8715 KB |

$mem \ cost$ = $mem\_{after} - mem\_{before} $
- $mem\_{before}$: before inserting data
- $mem\_{after}$: after inserting data and loading collection

I used following command, `insert_times` wii be increased if want to insert much more data. Python client version is `3.10`.
```bash
python collection_prepare.py --host=10.224.0.56 --nb=10000 --insert_times=10
```

## Question

1. For massive data, 100w and 500w entities, each entity will occupied approximately 5.0553 KB, 4.6139 KB separately. While for data scales of 10w and 50w, each single entity will occupied approximately 21.8429 KB, 7.5435 KB separately. The results are not proportional to the size of the data. **Is there any different memory increasing strategy for small data scale and massive data scale?**

**Comment by yhmo** [comment]
Memory usage is not only for pure data. Even your index is "FLAT", the server internally does a lot of things: compactions, interaction with etcd/pulsar, internal RPC calls, etc. All these actions require cpu memory.

**Reply by simonwei97** [reply]
Yeah. But I want to know why small data scale (10w, 50w) cost more mem than massive data scale (100w, 500w and so on) ? 

Memory consumption should be almost the same for different scale dataset is what I expected.

I suspect that whether when the amount of data is relatively small, the memory mallocation strategy is increased by 2 times or more at each time. While for massive data, the memory mallocation strategy is  increased by like 1/4 or any reasonable multiple num ( < 2) at each time.

Small data scale will consume much more memory than massive data scale for many operations what you mentioned. Can we put it this way ?

**Reply by xiaofan-luan** [reply, marked as answer]
Hi @simonwei97 
your test results seems to be solid.

As a system, milvus has a lot of overhead. The small the dataset, the more overhead there will be. 
In your case, I think each data takes around 4.5-4.6KB memory, and the rest memory is cost by overhead.

For a standalone cluster, I would recommend to reserve 4GB for the overhead, and rest of the data takes 4.5 kB memory . The larger the dataset, the smaller the overhead would be.
